comp/remote-config/rcclient/rcclient.go

```diff
@@ -104,25 +103,58 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig)
-	if len(mergedConfig.LogLevel) > 0 && mergedConfig.LogLevel != rc.configState.LatestLogLevel {
-		pkglog.Infof("Changing log level to %s through remote config", mergedConfig.LogLevel)
+	// Checks who (the source) is responsible for the last logLevel change
+	// The priority between sources is: CLI > RC > Default
+	source, err := settings.GetRuntimeSettingSource("log_level")
+	if err != nil {
+		pkglog.Errorf(err.Error())
+	}
+
+	switch source {
+	case settings.LogLevelSourceDefault:
+		// If the log level had been set by default
+		// and if we receive an empty value for log level in the config
+		// then there is nothing to do
+		if len(mergedConfig.LogLevel) == 0 {
+			return
+		}
+
 		// Get the current log level
-		var newFallback seelog.LogLevel
-		newFallback, err = pkglog.GetLogLevel()
-		if err == nil {
-			rc.configState.FallbackLogLevel = newFallback.String()
-			err = settings.SetRuntimeSetting("log_level", mergedConfig.LogLevel)
-			rc.configState.LatestLogLevel = mergedConfig.LogLevel
+		var newFallback interface{}
+		newFallback, err = settings.GetRuntimeSetting("log_level")
+		if err != nil {
+			break
 		}
-	} else {
-		var currentLogLevel seelog.LogLevel
-		currentLogLevel, err = pkglog.GetLogLevel()
-		if err == nil && currentLogLevel.String() == rc.configState.LatestLogLevel {
-			pkglog.Infof("Removing remote-config log level override, falling back to %s", rc.configState.FallbackLogLevel)
-			err = settings.SetRuntimeSetting("log_level", rc.configState.FallbackLogLevel)
+
+		pkglog.Infof("Changing log level to %s through remote config", mergedConfig.LogLevel)
+		rc.configState.FallbackLogLevel = newFallback.(string)
+		// Need to update the log level even if the level stays the same because we need to update the source
+		// Might be possible to add a check in deeper functions to avoid unnecessary work
+		err = settings.SetRuntimeSetting("log_level", mergedConfig.LogLevel, settings.LogLevelSourceRC)
+
+	case settings.LogLevelSourceRC:
+		// 2 possible situations:
+		//     - we want to change (once again) the log level through RC
+		//     - we want to fall back to the log level we had saved as fallback (in that case mergedConfig.LogLevel == "")
+		var newLevel string
+		var newSource settings.LogLevelSource
+		if len(mergedConfig.LogLevel) == 0 {
+			newLevel = rc.configState.FallbackLogLevel
+			newSource = settings.LogLevelSourceDefault
+			pkglog.Infof("Removing remote-config log level override, falling back to %s", newLevel)
+		} else {
+			newLevel = mergedConfig.LogLevel
+			newSource = settings.LogLevelSourceRC
+			pkglog.Infof("Changing log level to %s through remote config", newLevel)
 		}
+		err = settings.SetRuntimeSetting("log_level", newLevel, newSource)
+
+	case settings.LogLevelSourceCLI:
+		pkglog.Infof("Remote config could not change the log level due to CLI override")
+		return
+
+	default: // case settings.LogLevelSourceUnknown
+		pkglog.Warnf("Unknown source changed the log level")
 	}
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 Thought:** ‏This feels a bit bloated, with duplicated code between default/rc and cli/unknown. We just want to bail out early if:
- The last source was CLI
- We don't know what the last source was

IMO this does not warrant the whole switch, just an early return for these two conditions should be enough. WDYT?